File: system-test/java/src/test/java/com/optivem/eshop/systemtest/_legacy/mod06/base/BaseChannelDriverTest.java
package com.optivem.eshop.systemtest._legacy.mod06.base;

import com.optivem.eshop.systemtest.configuration.BaseConfigurableTest;
import com.optivem.eshop.systemtest.configuration.Configuration;
import com.optivem.eshop.dsl.driver.adapter.external.erp.ErpRealDriver;
import com.optivem.eshop.dsl.channel.ChannelType;
import com.optivem.eshop.dsl.driver.adapter.shop.api.ShopApiDriver;
import com.optivem.eshop.dsl.driver.port.shop.ShopDriver;
import com.optivem.eshop.dsl.driver.adapter.shop.ui.ShopUiDriver;
import com.optivem.eshop.systemtest.infrastructure.playwright.BrowserLifecycleExtension;
import com.optivem.eshop.dsl.common.Closer;
import com.optivem.testing.contexts.ChannelContext;
import com.optivem.testing.extensions.ChannelExtension;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(ChannelExtension.class)
public class BaseChannelDriverTest extends BaseConfigurableTest {
    protected ShopDriver shopDriver;
    protected ErpRealDriver erpDriver;

    @BeforeEach
    void setUp() {
        var configuration = loadConfiguration();

        shopDriver = createChannelShopDriver(configuration);
        erpDriver = new ErpRealDriver(configuration.getErpBaseUrl());
    }

    @AfterEach
    void tearDown() {
        Closer.close(shopDriver);
        Closer.close(erpDriver);
    }

    private ShopDriver createChannelShopDriver(Configuration configuration) {
        var channel = ChannelContext.get();

        if(channel == null) {
            return null;
        }

        if (ChannelType.UI.equals(channel)) {
            return new ShopUiDriver(configuration.getShopUiBaseUrl(), BrowserLifecycleExtension.getBrowser());
        } else if (ChannelType.API.equals(channel)) {
            return new ShopApiDriver(configuration.getShopApiBaseUrl());
        } else {
            throw new IllegalStateException("Unknown channel: " + channel);
        }
    }
}

File: system-test/dotnet/SystemTests/Legacy/Mod06/Base/BaseChannelDriverTest.cs
using SystemTests.TestInfrastructure.Configuration;
using Dsl.Core;
using Driver.Port.External.Erp;
using Driver.Adapter.Shop.Api;
using Driver.Adapter.Shop.Ui;
using Dsl.Core.Shop;
using Driver.Port.Shop;
using Driver.Adapter.External.Erp;
using Optivem.Testing;
using Xunit;

namespace SystemTests.Legacy.Mod06.Base;


public abstract class BaseChannelDriverTest : BaseConfigurableTest, IAsyncLifetime
{
    protected IShopDriver? _shopDriver;
    protected ErpRealDriver? _erpDriver;

    public virtual async Task InitializeAsync()
    {
        await TestLock.WaitAsync();
        await SetupDrivers();
    }

    private async Task SetupDrivers()
    {
        var configuration = LoadConfiguration();

        // Only create shop driver if channel context is set (for channel-parameterized tests)
        // For non-channel tests (like Erp), skip shop driver creation
        try
        {
            _shopDriver = await CreateShopDriverAsync(configuration);
        }
        catch (InvalidOperationException ex) when (ex.Message.Contains("Channel type is not set"))
        {
            _shopDriver = null;
        }

        _erpDriver = new ErpRealDriver(configuration.ErpBaseUrl);
    }

    public virtual async Task DisposeAsync()
    {
        try
        {
            if (_shopDriver != null)
                await _shopDriver.DisposeAsync();

            _erpDriver?.Dispose();
        }
        finally
        {
            TestLock.Release();
        }
    }

    protected async Task SetChannelAsync(Channel channel)
    {
        ChannelContext.Set(channel.Type);
        await SetupDrivers();
    }

    private static async Task<IShopDriver?> CreateShopDriverAsync(Dsl.Core.Configuration configuration)
    {
        var channelType = ChannelContext.Get();

        if (channelType == ChannelType.UI)
        {
            return await ShopUiDriver.CreateAsync(configuration.ShopUiBaseUrl);
        }
        else if (channelType == ChannelType.API)
        {
            return new ShopApiDriver(configuration.ShopApiBaseUrl);
        }
        else
        {
            throw new InvalidOperationException($"Unknown channel: {channelType}");
        }
    }
}













